--- node.h.orig	2018-09-13 16:43:52.114836010 -0700
+++ node.h	2018-09-13 16:45:50.076780325 -0700
@@ -157,6 +157,8 @@
    */
   typedef Callback<void,Ptr<NetDevice>, Ptr<const Packet>,uint16_t,const Address &,
                    const Address &, NetDevice::PacketType> ProtocolHandler;
+  typedef Callback<void,Ptr<NetDevice>, std::vector<uint8_t> & ,uint16_t,const Address &, const Address &, NetDevice::PacketType, bool> FateHandler;
+
   /**
    * \param handler the handler to register
    * \param protocolType the type of protocol this handler is 
@@ -174,6 +176,8 @@
                                 uint16_t protocolType,
                                 Ptr<NetDevice> device,
                                 bool promiscuous=false);
+  void RegisterFateProtocolHandler (FateHandler handler);
+
   /**
    * \param handler the handler to unregister
    *
@@ -287,7 +291,9 @@
   std::vector<Ptr<NetDevice> > m_devices; //!< Devices associated to this node
   std::vector<Ptr<Application> > m_applications; //!< Applications associated to this node
   ProtocolHandlerList m_handlers; //!< Protocol handlers in the node
+  FateHandler m_fateHandler; //!< Protocol handlers in the node
   DeviceAdditionListenerList m_deviceAdditionListeners; //!< Device addition listeners in the node
+  bool m_fateExists;
 };
 
 } // namespace ns3
